chain-signatures/node/src/protocol/presignature.rs

```diff
@@ -135,25 +225,131 @@ impl PresignatureManager {
+        self.presignature_storage
+            .write()
+            .await
+            .contains(id)
+            .map_err(|e| {
+                tracing::warn!(?e, "failed to check if presignature exist");
+            })
+            .unwrap_or(false)
+    }
+
+    /// Returns true if the mine presignature with the given id is already generated
+    pub async fn contains_mine(&self, id: &PresignatureId) -> bool {
+        self.presignature_storage
+            .write()
+            .await
+            .contains_mine(id)
+            .map_err(|e| {
+                tracing::warn!(?e, "failed to check if mine presignature exist");
+            })
+            .unwrap_or(false)
+    }
+
+    pub async fn take(&mut self, id: PresignatureId) -> Result<Presignature, GenerationError> {
+        if let Some(presignature) =
+            self.presignature_storage
+                .write()
+                .await
+                .take(&id)
+                .map_err(|e| {
+                    tracing::error!(?e, "failed to look for presignature");
+                    GenerationError::PresignatureIsMissing(id)
+                })?
+        {
+            self.gc.insert(id, Instant::now());
+            tracing::info!(id, "took presignature");
+            return Ok(presignature);
+        };
+
+        if self.generators.contains_key(&id) {
+            tracing::warn!(id, "presignature is still generating");
+            return Err(GenerationError::PresignatureIsGenerating(id));
+        }
+        if self.gc.contains_key(&id) {
+            tracing::warn!(id, "presignature was garbage collected");
+            return Err(GenerationError::PresignatureIsGarbageCollected(id));
+        }
+        tracing::warn!(id, "presignature is missing");
+        Err(GenerationError::PresignatureIsMissing(id))
+    }
+
+    pub async fn take_mine(&mut self) -> Option<Presignature> {
+        if let Some(presignature) = self
+            .presignature_storage
+            .write()
+            .await
+            .take_mine()
+            .map_err(|e| {
+                tracing::error!(?e, "failed to look for mine presignature");
+            })
+            .ok()?
+        {
+            tracing::info!(id = ?presignature.id, "took presignature of mine");
+            return Some(presignature);
+        }
+        None
+    }
+
     /// Returns the number of unspent presignatures available in the manager.
-    pub fn len(&self) -> usize {
-        self.presignatures.len()
+    pub async fn count_all(&self) -> usize {
+        self.presignature_storage
+            .write()
+            .await
+            .count_all()
+            .map_err(|e| {
+                tracing::error!(?e, "failed to count all presignatures");
+            })
+            .unwrap_or(0)
     }
 
     /// Returns the number of unspent presignatures assigned to this node.
-    pub fn my_len(&self) -> usize {
-        self.mine.len()
+    pub async fn count_mine(&self) -> usize {
```

Comment:
why change it all to `count_*`? Now it's gonna be inconsistent with what we have in triple manager

Reply:
I will make a followup right after this PR.